### PR TITLE
fix(dot/sync): fix `Timestamp slot must match 'CurrentSlot'` while using `westend` spec file

### PR DESCRIPTION
**Pull request by EclesioMeloJunior**
## Changes

Fixes the following tests:

- [x] TestChainProcessor_HandleBlockResponse_ValidChain
- [x] TestChainProcessor_HandleBlockResponse_MissingBlocks
- [x] TestChainProcessor_HandleBlockResponse_BlockData
- [x] TestChainProcessor_ExecuteBlock

## Tests

```sh
go test -timeout 10m github.com/ChainSafe/gossamer/dot/sync --tags=integration -v
```

## Issues

- Ref: https://github.com/chainsafe/gossamer/issues/3016
- Ref: https://github.com/ChainSafe/gossamer/issues/2895

## Primary Reviewer

@qdm12

### CODE REVIEW

File: dot/sync/chain_processor_integration_test.go [hunk 1]
```diff
@@ -13,14 +13,88 @@ import (
...
+		StateRoot:      common.Hash{},
+		ExtrinsicsRoot: common.Hash{},
+		Number:         parent.Number + 1,
+		Digest:         digest,
+	}
+
+	err = instance.InitializeBlock(header)
+	require.NoError(t, err)
+
+	idata := types.NewInherentData()
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:45]: nit `idata` -> `inherentData`

File: dot/sync/chain_processor_integration_test.go [hunk 2]
```diff
@@ -13,14 +13,88 @@ import (
...
+
+	ienc, err := idata.Encode()
+	require.NoError(t, err)
+
+	// Call BlockBuilder_inherent_extrinsics which returns the inherents as encoded extrinsics
+	inherentExts, err := instance.InherentExtrinsics(ienc)
+	require.NoError(t, err)
+
+	// decode inherent extrinsics
+	var inExts [][]byte
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:70]: nit no need for comment if the var is named explicitly imo

```suggestion
	var inherentExtrinsics [][]byte
```

File: dot/sync/chain_processor_integration_test.go [hunk 3]
```diff
@@ -13,14 +13,88 @@ import (
...
+	require.NoError(t, err)
+
+	err = idata.SetInherent(types.Newheads, []byte{0})
+	require.NoError(t, err)
+
+	ienc, err := idata.Encode()
+	require.NoError(t, err)
+
+	// Call BlockBuilder_inherent_extrinsics which returns the inherents as encoded extrinsics
+	inherentExts, err := instance.InherentExtrinsics(ienc)
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:66]: nit `inherentExts` -> `encodedInherentExtrinsics`

File: dot/sync/chain_processor_integration_test.go [hunk 4]
```diff
@@ -13,14 +13,88 @@ import (
...
+	require.NoError(t, err)
+
+	// decode inherent extrinsics
+	var inExts [][]byte
+	err = scale.Unmarshal(inherentExts, &inExts)
+	require.NoError(t, err)
+
+	// apply each inherent extrinsic
+	for _, inherent := range inExts {
+		in, err := scale.Marshal(inherent)
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:76]: nit `in` -> `encodedInherent`

File: dot/sync/chain_processor_integration_test.go [hunk 5]
```diff
@@ -13,14 +13,88 @@ import (
...
+		require.NoError(t, err)
+
+		ret, err := instance.ApplyExtrinsic(in)
+		require.NoError(t, err)
+		require.Equal(t, ret, []byte{0, 0})
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:81]: nit `ret` -> `result` or `applyExtrinsicResult` (per spec)

File: dot/sync/chain_processor_integration_test.go [hunk 6]
```diff
@@ -13,14 +13,88 @@ import (
...
+	}
+
+	res, err := instance.FinalizeBlock()
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:84]: nit `res` -> `finalisedHeader`

File: dot/sync/chain_processor_integration_test.go [hunk 7]
```diff
@@ -93,26 +178,52 @@ func TestChainProcessor_HandleBlockResponse_MissingBlocks(t *testing.T) {
 	rt, err := syncer.blockState.GetRuntime(bestBlockHash)
 	require.NoError(t, err)
 
+	babeCfg, err := rt.BabeConfiguration()
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:181]: supernit `babeCfg` -> `babeConfig`

File: dot/sync/chain_processor_integration_test.go [hunk 8]
```diff
@@ -196,7 +311,15 @@ func TestChainProcessor_ExecuteBlock(t *testing.T) {
...
 	require.NoError(t, err)
 
-	block := BuildBlock(t, rt, parent, nil)
+	babeCfg, err := rt.BabeConfiguration()
+	require.NoError(t, err)
+
+	timestamp := uint64(time.Now().Unix())
+	slotDuration := babeCfg.SlotDuration
+
+	// calcule the exact slot for each produced block
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:320]:
```suggestion
	// calculate the exact slot for each produced block
```

File: dot/sync/chain_processor_integration_test.go [hunk 9]
```diff
@@ -164,7 +273,15 @@ func TestChainProcessor_HandleBlockResponse_BlockData(t *testing.T) {
...
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:282]:
```suggestion
	// calculate the exact slot for each produced block
```

File: dot/sync/chain_processor_integration_test.go [hunk 10]
```diff
@@ -93,26 +178,52 @@ func TestChainProcessor_HandleBlockResponse_MissingBlocks(t *testing.T) {
...
 
+	babeCfg, err = rt.BabeConfiguration()
+	require.NoError(t, err)
+
+	timestamp = uint64(time.Now().Unix())
+	slotDuration = babeCfg.SlotDuration
+
 	for i := 0; i < 16; i++ {
-		block := BuildBlock(t, rt, parent, nil)
+		// calcule the exact slot for each produced block
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:216]:
```suggestion
		// calculate the exact slot for each produced block
```

File: dot/sync/chain_processor_integration_test.go [hunk 11]
```diff
@@ -93,26 +178,52 @@ func TestChainProcessor_HandleBlockResponse_MissingBlocks(t *testing.T) {
...
 
+	babeCfg, err := rt.BabeConfiguration()
+	require.NoError(t, err)
+
+	timestamp := uint64(time.Now().Unix())
+	slotDuration := babeCfg.SlotDuration
+
 	for i := 0; i < 4; i++ {
-		block := BuildBlock(t, rt, parent, nil)
+		// calcule the exact slot for each produced block
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:188]:
```suggestion
		// calculate the exact slot for each produced block
```

File: dot/sync/chain_processor_integration_test.go [hunk 12]
```diff
@@ -33,11 +107,24 @@ func TestChainProcessor_HandleBlockResponse_ValidChain(t *testing.T) {
...
 
+	babeCfg, err := rt.BabeConfiguration()
+	require.NoError(t, err)
+
+	timestamp := uint64(time.Now().Unix())
+	slotDuration := babeCfg.SlotDuration
+
 	for i := 0; i < maxResponseSize*2; i++ {
-		block := BuildBlock(t, rt, parent, nil)
+		// calcule the exact slot for each produced block
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:117]:
```suggestion
		// calculate the exact slot for each produced block
```

File: dot/sync/chain_processor_integration_test.go [hunk 13]
```diff
@@ -33,11 +107,24 @@ func TestChainProcessor_HandleBlockResponse_ValidChain(t *testing.T) {
...
+
+		block := buildBlockWithSlotAndTimestamp(t, rt, parent, currentSlot, timestamp)
 		err = responder.blockState.(*state.BlockState).AddBlock(block)
 		require.NoError(t, err)
 		parent = &block.Header
+
+		// increase the timestamp by the slot duration
+		// so we will get a different slot for the next block
+		timestamp += slotDuration
 	}
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:128]: Should we move the for loop inner body to a function? It looks repeated a bunch of times in this file

**Review by qdm12 — APPROVED** [review]

File: dot/sync/chain_processor_integration_test.go [hunk 14]
```diff
@@ -13,33 +13,127 @@ import (
...
+	err = inherentData.SetInherent(types.Parachn0, parachainInherent)
+	require.NoError(t, err)
+
+	err = inherentData.SetInherent(types.Newheads, []byte{0})
 	require.NoError(t, err)
 
-	for i := 0; i < maxResponseSize*2; i++ {
-		block := BuildBlock(t, rt, parent, nil)
-		err = responder.blockState.(*state.BlockState).AddBlock(block)
+	ienc, err := inherentData.Encode()
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:62]: nit `ienc` -> `encodedInherentData`

File: dot/sync/chain_processor_integration_test.go [hunk 15]
```diff
@@ -13,33 +13,127 @@ import (
...
+	finalisedHeader.Number = header.Number
+	finalisedHeader.Hash()
+
+	return &types.Block{
+		Header: *finalisedHeader,
+		Body:   body,
+	}
+}
+
+func buildAndAddBlocksToState(t *testing.T, runtime state.Runtime, blockState *state.BlockState, amount int) {
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:98]: supernit make `amount` an `uint` (aka restrict dev error as much as possible 😄)

```suggestion
func buildAndAddBlocksToState(t *testing.T, runtime state.Runtime, blockState *state.BlockState, amount uint) {
```

File: dot/sync/chain_processor_integration_test.go [hunk 16]
```diff
@@ -13,33 +13,127 @@ import (
...
 	}
 
+}
+
+func TestChainProcessor_HandleBlockResponse_ValidChain(t *testing.T) {
+	syncer := newTestSyncer(t)
+	responder := newTestSyncer(t)
+
+	bestBlockHash := responder.blockState.(*state.BlockState).BestBlockHash()
+	rt, err := responder.blockState.GetRuntime(bestBlockHash)
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:131]: nit `rt` -> `runtime` / `runtimeInstance`

File: dot/sync/chain_processor_integration_test.go [hunk 17]
```diff
@@ -81,39 +175,22 @@ func TestChainProcessor_HandleBlockResponse_ValidChain(t *testing.T) {
...
+	syncerRuntime, err := syncer.blockState.GetRuntime(bestBlockHash)
 	require.NoError(t, err)
 
-	for i := 0; i < 4; i++ {
-		block := BuildBlock(t, rt, parent, nil)
-		err = syncer.blockState.(*state.BlockState).AddBlock(block)
-		require.NoError(t, err)
-		parent = &block.Header
-	}
+	const syncerAmoutOfBlocks = 4
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:185]:
```suggestion
	const syncerAmountOfBlocks = 4
```

File: dot/sync/chain_processor_integration_test.go [hunk 18]
```diff
@@ -81,39 +175,22 @@ func TestChainProcessor_HandleBlockResponse_ValidChain(t *testing.T) {
...
+	responderRuntime, err := responder.blockState.GetRuntime(bestBlockHash)
 	require.NoError(t, err)
 
-	for i := 0; i < 16; i++ {
-		block := BuildBlock(t, rt, parent, nil)
-		err = responder.blockState.(*state.BlockState).AddBlock(block)
-		require.NoError(t, err)
-		parent = &block.Header
-	}
+	const responderAmoutOfBlocks = 16
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:192]:
```suggestion
	const responderAmountOfBlocks = 16
```

File: dot/sync/chain_processor_integration_test.go [hunk 19]
```diff
@@ -13,33 +13,127 @@ import (
...
+	encodedInherentData, err := inherentData.Encode()
+	require.NoError(t, err)
+
+	// Call BlockBuilder_inherent_extrinsics which returns the inherents as encoded extrinsics
+	encodedInherentExtrinsics, err := instance.InherentExtrinsics(encodedInherentData)
+	require.NoError(t, err)
+
+	// decode inherent extrinsics
+	var inherentExtrinsics [][]byte
+	err = scale.Unmarshal(encodedInherentExtrinsics, &inherentExtrinsics)
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:71]: nit unneeded comment

```suggestion
	var inherentExtrinsics [][]byte
	err = scale.Unmarshal(encodedInherentExtrinsics, &inherentExtrinsics)
```

File: dot/sync/chain_processor_integration_test.go [hunk 20]
```diff
@@ -13,33 +13,127 @@ import (
...
+	require.NoError(t, err)
+
+	// apply each inherent extrinsic
+	for _, inherent := range inherentExtrinsics {
```
Comment (qdm12) [inline dot/sync/chain_processor_integration_test.go:75]: nit unneeded comment

```suggestion
	for _, inherent := range inherentExtrinsics {
```

**Review by qdm12 — APPROVED** [review]

**Review by kishansagathiya — APPROVED** [review]

**Review by jimjbrettj — APPROVED** [review]
nice work!

**Pull request merged**